## Project - 1 : Data Cleaning & Preparation

### Goal:
- Clean a raw dataset by handling missing values, duplicates, and incorrect data.
- Key Requirements:
  - Identify missing or null values
  -  Remove duplicates
  -  Correct data formats (dates,numbers, text)

#### Importing the dataset

In [3]:
import pandas as pd
import numpy as np

df = pd.read_excel('Dataset for Data Analytics.xlsx')
df

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1195,ORD201195,2024-06-20,C21126,Desk,1,107.04,392 Main St,Credit Card,Cancelled,TRK38009181,6,FREESHIP,Google,107.04
1196,ORD201196,2024-03-04,C20095,Monitor,2,662.53,778 Main St,Online,Cancelled,TRK69207593,5,NaN,Facebook,1325.06
1197,ORD201197,2023-07-13,C79674,Tablet,2,436.84,275 Main St,Online,Delivered,TRK88039356,2,FREESHIP,Instagram,873.68
1198,ORD201198,2024-08-22,C64753,Chair,4,262.52,509 Main St,Debit Card,Cancelled,TRK71683331,4,WINTER15,Instagram,1050.08


In [9]:
print('--Nulls--')
print(df.isnull().sum())

print('\n--dtypes--')
print(df.dtypes)

print('\n--Duplicates--')
print('Full row duplicates : ', df.duplicated().sum())
print('OrderID duplicates:', df.duplicated(subset=['OrderID']).sum())

--Nulls--
OrderID              0
Date                 0
CustomerID           0
Product              0
Quantity             0
UnitPrice            0
ShippingAddress      0
PaymentMethod        0
OrderStatus          0
TrackingNumber       0
ItemsInCart          0
CouponCode         309
ReferralSource       0
TotalPrice           0
dtype: int64

--dtypes--
OrderID                    object
Date               datetime64[ns]
CustomerID                 object
Product                    object
Quantity                    int64
UnitPrice                 float64
ShippingAddress            object
PaymentMethod              object
OrderStatus                object
TrackingNumber             object
ItemsInCart                 int64
CouponCode                 object
ReferralSource             object
TotalPrice                float64
dtype: object

--Duplicates--
Full row duplicates :  0
OrderID duplicates: 0


### Observations from the dataset : 

- No duplicate OrderIDs
- No full row duplicates
- Dates already clean
- 309 null values in CouponCode (Cleaning work required)
- Both Categorical and Numerical columns looks good

#### Phase 1 : Strategic Imputation (handling the gaps) 

- CouponCode nulls doesn't mean 'missing data', this mean the customers used No coupon. So filling it with 'No Coupon' rather than deleting.

In [11]:
df["CouponCode"] = df["CouponCode"].fillna('No Coupon')

df.isnull().sum()

OrderID            0
Date               0
CustomerID         0
Product            0
Quantity           0
UnitPrice          0
ShippingAddress    0
PaymentMethod      0
OrderStatus        0
TrackingNumber     0
ItemsInCart        0
CouponCode         0
ReferralSource     0
TotalPrice         0
dtype: int64

#### Phase 2 : Integrity Audit (checking unique IDs)

In [13]:
duplics = df[df.duplicated(subset=['OrderID'])]
print(f'Duplicated OrderIDs : {len(duplics)}')

Duplicated OrderIDs : 0


In [14]:
#if duplicated OrderIDs were there then :

if len(duplics) > 0:
    df = df.drop_duplicates(subset = ['OrderID'])
    print(f'Rows After duplicates removed : {len(duplics)}')
else:
    print('All OrderIDs are unique!')

All OrderIDs are unique!


#### Phase 3 : Standerdize (speak one language)

In [25]:
#proper text casing on some cols:

for col in ['Product', 'PaymentMethod', 'OrderStatus', 'ReferralSource']:
    df[col] = df[col].str.strip().str.title()

# df['Product'].unique()


#date to ISO 8601 (YYYY-MM-DD)
df['Date'] = pd.to_datetime(df['Date'].dt.strftime('%Y-%m-%d'))


#round off numericals to 2 decimals:
df['UnitPrice'] = df['UnitPrice'].round(2)
df['TotalPrice'] = df['TotalPrice'].round(2)

print('After Standerdization : ')
df.sample()

After Standerdization : 


,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
404,ORD200404,2023-03-09,C45482,Monitor,2,99.1,440 Main St,Cash,Delivered,TRK54571699,6,No Coupon,Instagram,198.2


#### Threshold for Project 2 (final verification gate)

In [28]:
print('--Final Audit--')
print('Total rows : ', len(df))
print('Nulls remaining : ', df.isnull().sum().sum())
print('Duplicate OrderIDs : ', df.duplicated(subset='OrderID').sum())

--Final Audit--
Total rows :  1200
Nulls remaining :  0
Duplicate OrderIDs :  0


#### Saving the cleaned file :

In [29]:
df.to_excel('Dataset_Cleaned.xlsx', index=False)
print('Saved!')

Saved!
